### PR TITLE
Cleaned LightGBM documentation

### PR DESCRIPTION
**Pull request by shmoradims**
LightGBM API, trainers, boosters, and options documentation. Part of #2522.
Source of documentation are:
* https://github.com/Microsoft/LightGBM/blob/master/docs/Parameters.rst
* https://lightgbm.readthedocs.io/en/latest/Parameters.html
* ML.NET source code

### CODE REVIEW

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs [hunk 1]
```diff
@@ -129,54 +132,114 @@ internal static class Defaults
             public const int NumberOfIterations = 100;
         }
 
+        /// <summary>
+        /// Gradient boosting decision tree.
+        /// </summary>
+        /// <remarks>
+        /// For details, please see <a href="https://en.wikipedia.org/wiki/Gradient_boosting#Gradient_tree_boosting">here</a>.
```
Comment (sfilipi) [inline src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: replace with "Gradient_tree_boosting"

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs [hunk 2]
```diff
@@ -129,54 +132,114 @@ internal static class Defaults
...
                         "this simply corresponds to minimum number of instances needed to be in each node. The larger, the more conservative the algorithm will be.")]
                 [TlcModule.Range(Min = 0.0)]
                 public double MinimumChildWeight = 0.1;
 
+                /// <summary>
+                /// The frequency of performing subsamplig (bagging).
+                /// </summary>
+                /// <value>
+                /// 0 means disable bagging. N means perform bagging at every N iterations.
+                /// To enable bagging, <see cref="SubsampleFraction"/> should also be set to a value less than 1.0.
```
Comment (sfilipi) [inline src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: >should also be set to a value less than 1.0 [](start = 71, length = 43)

Just curious, where did this information come from?

File: src/Microsoft.ML.LightGBM/LightGbmCatalog.cs [hunk 3]
```diff
@@ -46,14 +46,14 @@ public static class LightGbmExtensions
         }
 
         /// <summary>
-        /// Predict a target using a decision tree regression model trained with the <see cref="LightGbmRegressorTrainer"/>.
+        /// Predict a target using a boosted decision tree regression model trained with the <see cref="LightGbmRegressorTrainer"/> and advanced options.
         /// </summary>
         /// <param name="catalog">The <see cref="RegressionCatalog"/>.</param>
-        /// <param name="options">Advanced options to the algorithm.</param>
+        /// <param name="options">Trainer options.</param>
```
Comment (sfilipi) [inline src/Microsoft.ML.LightGBM/LightGbmCatalog.cs]: >Trainer [](start = 34, length = 7)

either estimator or algorithm

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs (hunk 1 above)
Comment (shmoradims) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: done

---
In reply to: [263686400](https://github.com/dotnet/machinelearning/pull/2886#discussion_r263686400) [](ancestors = 263686400)

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs (hunk 2 above)
Comment (shmoradims) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: three sources: 
*[https://github.com/Microsoft/LightGBM/blob/master/docs/Parameters.rst](https://github.com/Microsoft/LightGBM/blob/master/docs/Parameters.rst)
*[https://lightgbm.readthedocs.io/en/latest/Parameters.html](https://lightgbm.readthedocs.io/en/latest/Parameters.html)
*ML.NET source code

---
In reply to: [263689038](https://github.com/dotnet/machinelearning/pull/2886#discussion_r263689038) [](ancestors = 263689038)

File: src/Microsoft.ML.LightGBM/LightGbmCatalog.cs (hunk 3 above)
Comment (shmoradims) [reply src/Microsoft.ML.LightGBM/LightGbmCatalog.cs]: I've been changing them to "Trainer Options" based on Rogan's comment on AveragedPercentron. I don't want to change back everything. Trainer sounds good enough,  unless there's a strong objection. 

---
In reply to: [263689283](https://github.com/dotnet/machinelearning/pull/2886#discussion_r263689283) [](ancestors = 263689283)

File: test/BaselineOutput/Common/EntryPoints/core_manifest.json [hunk 4]
```diff
@@ -11360,7 +11360,7 @@
         {
           "Name": "CustomGains",
           "Type": "String",
-          "Desc": "Comma seperated list of gains associated to each relevance label.",
+          "Desc": "Comma separated list of gains associated to each relevance label.",
```
Comment (rogancarr) [inline test/BaselineOutput/Common/EntryPoints/core_manifest.json]: >separated [](start = 25, length = 9)

Nice

File: src/Microsoft.ML.LightGBM/doc.xml [hunk 5]
```diff
@@ -0,0 +1,15 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<doc>
+  <members>
+    <!--  
+    The following text describes the LightGbm algorithm details.
+    It's used for the remarks section of all LightGbm-based trainers (binary, multiclass, regression, ranking)
+    -->
+    <member name="LightGBM_remarks">
+      <remarks>
+        LightGBM is an open source implementation of boosted trees. For implementation details, please see
```
Comment (rogancarr) [inline src/Microsoft.ML.LightGBM/doc.xml]: >boosted trees [](start = 53, length = 13)

gradient boosted decision trees. (or Gradient Boosting Machine for GBM?)

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs [hunk 6]
```diff
@@ -38,9 +38,12 @@ public interface IBoosterParameter
...
 
     /// <summary>
-    /// Parameters names comes from LightGBM library.
-    /// See https://github.com/Microsoft/LightGBM/blob/master/docs/Parameters.rst.
+    /// Options for LightGBM trainer.
     /// </summary>
+    /// <remarks>
+    /// LightGBM is an external library that's integrated with ML.NET. For detailed information about the parameters
+    /// please see https://github.com/Microsoft/LightGBM/blob/master/docs/Parameters.rst.
+    /// </remarks>
```
Comment (rogancarr) [inline src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: Is there a version info anywhere that we could put here so that it's clear what version we have and it's auto-updated when we update LightGBM?

File: src/Microsoft.ML.LightGBM/doc.xml [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+<doc>
+  <members>
+    <!--  
+    The following text describes the LightGbm algorithm details.
+    It's used for the remarks section of all LightGbm-based trainers (binary, multiclass, regression, ranking)
+    -->
+    <member name="LightGBM_remarks">
+      <remarks>
+        LightGBM is an open source implementation of boosted trees. For implementation details, please see 
+        <a href='https://lightgbm.readthedocs.io/en/latest/index.html'>LightGBM's official documentation</a>.
```
Comment (rogancarr) [inline src/Microsoft.ML.LightGBM/doc.xml]: >documentation [](start = 91, length = 13)

Also link to the paper with Guolin and Tom.

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs [hunk 8]
```diff
@@ -314,98 +431,184 @@ public enum EvalMetricType
...
+        /// <value>
+        /// 0 means disable early stopping.
+        /// </value>
         [Argument(ArgumentType.AtMostOnce, HelpText = "Rounds of early stopping, 0 will disable it.",
             ShortName = "es")]
         public int EarlyStoppingRound = 0;
 
-        [Argument(ArgumentType.AtMostOnce, HelpText = "Comma seperated list of gains associated to each relevance label.", ShortName = "gains")]
+        /// <summary>
+        /// Comma separated list of gains associated with each relevance label. Used only by <see cref="LightGbmRankingTrainer"/>.
```
Comment (rogancarr) [inline src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: >a sep [](start = 16, length = 5)

hyphen

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs [hunk 9]
```diff
@@ -314,98 +431,184 @@ public enum EvalMetricType
...
+        /// <summary>
+        /// L2 regularization for categorical split.
+        /// </summary>
         [Argument(ArgumentType.AtMostOnce, HelpText = "L2 Regularization for categorical split.")]
         [TlcModule.Range(Min = 0.0)]
         [TlcModule.SweepableDiscreteParam("CatL2", new object[] { 0.1, 0.5, 1, 5, 10 })]
         public double L2CategoricalRegularization = 10;
 
+        /// <summary>
+        /// The random seed for LightGBM to use.
```
Comment (rogancarr) [inline src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: >The random seed for LightGBM to use. [](start = 12, length = 36)

Please specify what happens if the seed is not set by the user.

File: src/Microsoft.ML.LightGBM/LightGbmCatalog.cs [hunk 10]
```diff
@@ -14,20 +14,20 @@ namespace Microsoft.ML
     public static class LightGbmExtensions
     {
         /// <summary>
-        /// Predict a target using a decision tree regression model trained with the <see cref="LightGbmRegressorTrainer"/>.
+        /// Predict a target using a boosted decision tree regression model trained with the <see cref="LightGbmRegressorTrainer"/>.
```
Comment (rogancarr) [inline src/Microsoft.ML.LightGBM/LightGbmCatalog.cs]: > a boosted decision tree re [](start = 34, length = 27)

"gradient boosting decision tree" to be consistent with other docs.

File: src/Microsoft.ML.LightGBM/LightGbmCatalog.cs [hunk 11]
```diff
@@ -46,14 +46,14 @@ public static class LightGbmExtensions
         }
 
         /// <summary>
-        /// Predict a target using a decision tree regression model trained with the <see cref="LightGbmRegressorTrainer"/>.
+        /// Predict a target using a boosted decision tree regression model trained with the <see cref="LightGbmRegressorTrainer"/> and advanced options.
```
Comment (rogancarr) [inline src/Microsoft.ML.LightGBM/LightGbmCatalog.cs]: >oosted decision tree regression [](start = 38, length = 31)

Ditto

File: src/Microsoft.ML.LightGBM/LightGbmCatalog.cs (hunk 3 above)
Comment (rogancarr) [reply src/Microsoft.ML.LightGBM/LightGbmCatalog.cs]: I think this is preferred to "advanced".

---
In reply to: [263905459](https://github.com/dotnet/machinelearning/pull/2886#discussion_r263905459) [](ancestors = 263905459,263689283)

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs [hunk 12]
```diff
@@ -129,54 +132,114 @@ internal static class Defaults
...
+                /// </value>
                 [Argument(ArgumentType.AtMostOnce,
                     HelpText = "Minimum sum of instance weight(hessian) needed in a child. If the tree partition step results in a leaf " +
                         "node with the sum of instance weight less than min_child_weight, then the building process will give up further partitioning. In linear regression mode, " +
                         "this simply corresponds to minimum number of instances needed to be in each node. The larger, the more conservative the algorithm will be.")]
                 [TlcModule.Range(Min = 0.0)]
                 public double MinimumChildWeight = 0.1;
 
+                /// <summary>
+                /// The frequency of performing subsamplig (bagging).
```
Comment (zeahmed) [inline src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: >subsamplig [](start = 48, length = 10)

subsampling.

File: src/Microsoft.ML.LightGBM/doc.xml (hunk 5 above)
Comment (shmoradims) [reply src/Microsoft.ML.LightGBM/doc.xml]: changed to "Gradient Boosting Decision Tree" based on Tom's paper title.

---
In reply to: [263944693](https://github.com/dotnet/machinelearning/pull/2886#discussion_r263944693) [](ancestors = 263944693)

File: src/Microsoft.ML.LightGBM/doc.xml (hunk 7 above)
Comment (shmoradims) [reply src/Microsoft.ML.LightGBM/doc.xml]: good point.

---
In reply to: [263945392](https://github.com/dotnet/machinelearning/pull/2886#discussion_r263945392) [](ancestors = 263945392)

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs (hunk 8 above)
Comment (shmoradims) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: done

---
In reply to: [263946259](https://github.com/dotnet/machinelearning/pull/2886#discussion_r263946259) [](ancestors = 263946259)

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs (hunk 12 above)
Comment (shmoradims) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: good catch.

---
In reply to: [263963937](https://github.com/dotnet/machinelearning/pull/2886#discussion_r263963937) [](ancestors = 263963937)

File: src/Microsoft.ML.LightGBM/LightGbmCatalog.cs (hunk 11 above)
Comment (shmoradims) [reply src/Microsoft.ML.LightGBM/LightGbmCatalog.cs]: done

---
In reply to: [263946785](https://github.com/dotnet/machinelearning/pull/2886#discussion_r263946785) [](ancestors = 263946785)

File: src/Microsoft.ML.LightGBM/LightGbmCatalog.cs (hunk 10 above)
Comment (shmoradims) [reply src/Microsoft.ML.LightGBM/LightGbmCatalog.cs]: done

---
In reply to: [263946628](https://github.com/dotnet/machinelearning/pull/2886#discussion_r263946628) [](ancestors = 263946628)

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs (hunk 9 above)
Comment (shmoradims) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: done

---
In reply to: [263946446](https://github.com/dotnet/machinelearning/pull/2886#discussion_r263946446) [](ancestors = 263946446)

File: src/Microsoft.ML.LightGBM/LightGbmArguments.cs (hunk 6 above)
Comment (shmoradims) [reply src/Microsoft.ML.LightGBM/LightGbmArguments.cs]: I don't think we have an auto-update mechanism for external repo version in docs. Right now it's at 2.2.3 which is the latest. I can't think of a way to force-update the docs here, if LightGBM version is included in text.

It's also easy to look at the LightGBM nuget version that comes with ML.NET nugets. That's the most reliable way to find the version I think.


---
In reply to: [263945106](https://github.com/dotnet/machinelearning/pull/2886#discussion_r263945106) [](ancestors = 263945106)
